### PR TITLE
ENH: add Relu6GradGrad

### LINKED ISSUE #13144: No gradient defined for Relu6Grad
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes. MWE below is not a stock example.

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Linux Ubuntu 16.04

- **TensorFlow installed from (source or binary)**:
Binary (Miniconda)

- **TensorFlow version (use command below)**:
print(tf.GIT_VERSION, tf.VERSION)
('v1.2.0-5-g435cdfc', '1.2.1')

- **Python version**: 
2.7.13

- **Bazel version (if compiling from source)**:
N/A

- **CUDA/cuDNN version**:
CUDA 8.0.61
CuDNN 5.1.10

- **GPU model and memory**:
GeForce GTX TITAN X
Total Memory 11.91GiB

- **Exact command to reproduce**:
See below.

### Describe the problem

Describe the problem clearly here. Be sure to convey here why it's a bug in TensorFlow or a feature request.

One can not apply the `gradient` operator to a `nn.Relu6` more than once.
Applying it once gives a `Relu6Grad` node,
applying it a second time (i.e applying it to the `Relu6Grad`)  throws an error.

Unless I have messed up my math (quiet possible),
Relu6 is infinitely differentiable, except at 2 points,
albeit very boring

 - f(x) = relu6(x) 
 - df/dx = 1 if 0<x<6 else 0
 - d2f/dxx = 0
 - all further derivatives also 0.

How-ever, it is marginally more interesting if one is applying the chain rule to it

 - f(x1,x2) = relu6(x1*x2) 
 - df/dx1 = x2 if 0<x1*x2<6 else 0
 - d2f/dx1x2 = 1 if 0<x1*x2<6 else 0

This is a feature request to make the gradient of `Relu6Grad` defined.

### Source code / logs

MWE:
```python

import tensorflow as tf
sess = tf.Session()

x1 = tf.placeholder(tf.float32)
x2 = tf.placeholder(tf.float32)

y = tf.nn.relu6(x1*x2)
d1 = tf.gradients(y,x1)
d2 = tf.gradients(d1, x2)
```

Errors on the final line  with
```
---------------------------------------------------------------------------
LookupError                               Traceback (most recent call last)
<ipython-input-7-4250ab355a8b> in <module>()
----> 1 d2 = tf.gradients(d1, x2)

...usr/lib/python2.7/site-packages/te
nsorflow/python/ops/gradients_impl.pyc in gradients(ys, xs, grad_ys, name, colocate_gradients_with_ops,
 gate_gradients, aggregation_method)
    512               raise LookupError(
    513                   "No gradient defined for operation '%s' (op type: %s)" %
--> 514                   (op.name, op.type))
    515         if loop_state:
    516           loop_state.EnterGradWhileContext(op, before=False)

LookupError: No gradient defined for operation 'gradients/Relu6_grad/Relu6Grad' (op type: Relu6Grad)
```

### PR DESCRIPTION
**Pull request by facaiy**
Fix #13144.

### CODE REVIEW

**Comment by alextp** [comment]
Can you add a gradient_checker test (similar to other grad_grad functions) to verify that this works now and continues to work in the future?

**Comment by facaiy** [comment]
@alextp Sure! Since I'm on holiday, I'll add the test later.

**Comment by facaiy** [comment]
Hi, @alextp .  I am a little confusing with xxxGradGrad functions.

Say, a simple network as below:
L = h(y), 
y = func(x), 

### first-order gradient

As chain ruler, It is easy to get first-order gradient:
dL / dx = (dy / dx) * (dL / dy) 

And the implementation is straight as well:
```python
def xxxGrad(op, grad):
      x = op.inputs[0]
      y = op.outputs[0]
      dy_dx =  # math calculation
      return dy_dx * grad
```
Here `grad` is `dL / dy`.

### second-order gradient

However, for second-order gradient,  it seems that `xxxGradGrad` has two inputs.

Take `_SoftplusGradGrad` for example:

```python
def _SoftplusGradGrad(op, grad):
  # Let:
  #   y = tf.nn.softplus(x)
  #   dx = gen_nn_ops._softplus_grad(dy, x) = dy / (1 + exp(-x))
  # This op computes (ddy, d2x) from op.inputs == [dy, x] and grad == ddx.
  dy, x = op.inputs
  with ops.control_dependencies([grad.op]):
    ddy = gen_nn_ops._softplus_grad(grad, x)  # pylint: disable=protected-access
    d2x = grad * dy / (math_ops.exp(-x) + 2.0 + math_ops.exp(x))
return (ddy, d2x)
```
https://github.com/facaiy/tensorflow/blob/ed66e5a9dee770492bfdfe37d1a01c87d3bca3de/tensorflow/python/ops/nn_grad.py#L377

I cannot figure out why two inputs exist, and could you explain what are `dy`, `grad`, `ddy` and `d2x` here? Thank you very much.

**Comment by facaiy** [comment]
Oh, perhaps I understand it.

For `_SoftplusGradGrad`,
Let: 
y = tf.nn.softplus(x)
dx = gen_nn_ops._softplus_grad(dy, x)

here `dy` is short for `dL/dy`, `dx` is short for `dL/dx`. Since `dx` is a function with `dy` and `x`, so `_SoftplusGradGrad` need to calculate the gradient for `dy` and `x` respectively:
`ddy` is short for `d(dx) / dy`,
`d2x` is for `d(dx) / dx`.

right?

**Comment by facaiy** [comment]
Hi, test case has been added, however, something seems missing:

```python
======================================================================
ERROR: testRelu6GradGrad (__main__.Relu6OpTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data1/users/facai/.cache/bazel/_bazel_facai/3338df3cdc4fd0e5fdd3f3ae6490e0be/execroot/org_tensorflow/bazel-out/local-opt/bin/tensorflow/python/nn_grad_test.runfiles/org_tensorflow/tensorflow/python/ops/nn_grad_test.py", line 34, in testRelu6GradGrad
    r_g = gradients_impl.gradients(r, inputs)
  File "/data1/users/facai/.cache/bazel/_bazel_facai/3338df3cdc4fd0e5fdd3f3ae6490e0be/execroot/org_tensorflow/bazel-out/local-opt/bin/tensorflow/python/nn_grad_test.runfiles/org_tensorflow/tensorflow/python/ops/gradients_impl.py", line 516, in gradients
    (op.name, op.type))
LookupError: No gradient defined for operation 'Relu6' (op type: Relu6)
```

File: tensorflow/python/ops/nn_grad_test.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.python.framework import constant_op
+from tensorflow.python.framework import dtypes
+from tensorflow.python.ops import gradient_checker
+from tensorflow.python.ops import gradients_impl
+from tensorflow.python.ops import nn_ops
+from tensorflow.python.platform import test
+
```
Comment (alextp) [inline tensorflow/python/ops/nn_grad_test.py]: This file needs to import nn_grad to make the error about relu6 not having a gradient disappear

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/nn_grad_test.py (hunk 1 above)
Comment (facaiy) [reply tensorflow/python/ops/nn_grad_test.py]: Thank you, test modified as below and passed.

**Review by alextp — APPROVED** [review]

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by facaiy** [comment]
It seems unrelated failure

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by facaiy** [comment]
Thank you, @alextp . Latest master is merged, could we retest it?

**Comment by alextp** [comment]
Jenkins, test this please.

**Comment by vrv** [comment]
@tensorflow-jenkins test this please

**Comment by facaiy** [comment]
Hi, does anyone know what is wrong with Ubuntu CC? Thanks.

**Comment by vrv** [comment]
We have to kick it off manually using a separate label, just did that.

**Comment by facaiy** [comment]
@vrv thank you. Two tests time out, and it seems unrelated.

**Comment by vrv** [comment]
Yeah, unrelated, thanks!

**Pull request merged**